lib/at_cmd_parser/at_cmd_parser.c

```diff
@@ -0,0 +1,340 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+#include <ctype.h>
+#include <limits.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <string.h>
+#include <zephyr.h>
+#include <zephyr/types.h>
+
+#include <at_cmd_parser/at_cmd_parser.h>
+#include <at_cmd_parser/at_utils.h>
+
+#define AT_PARAM_SEPARATOR              ','
+#define AT_CMD_SEPARATOR                ':'
+#define AT_CMD_BUFFER_TERMINATOR        0
+#define AT_CMD_STRING_IDENTIFIER        '\"'
+#define AT_STANDARD_NOTIFICATION_PREFIX '+'
+#define AT_PROP_NOTIFICATION_PREFX      '%'
+
+enum at_parser_state {
+	DETECT_TYPE,
+	SEARCH_NEXT_PARAM,
+	ARRAY,
+	STRING,
+	NUMBER,
+	SMS_PDU,
+	NOTIFICATION,
+	OPTIONAL,
+};
+
+static enum at_parser_state state;
+static enum at_parser_state prev_state;
+
+static inline bool is_notification(char chr)
+{
+	if ((chr == AT_STANDARD_NOTIFICATION_PREFIX) ||
+	    (chr == AT_PROP_NOTIFICATION_PREFX)) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_valid_notification_char(char chr)
+{
+	chr = toupper(chr);
+
+	if ((chr >= 'A') && (chr <= 'Z')) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_terminated(char chr)
+{
+	if (chr == AT_CMD_BUFFER_TERMINATOR) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_separator(char chr)
+{
+	if ((chr == AT_PARAM_SEPARATOR) ||
+	    (chr == AT_CMD_SEPARATOR)) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_lfcr(char chr)
+{
+	if ((chr == '\r') ||
+	    (chr == '\n')) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_dblquote(char chr)
+{
+	if (chr == '"') {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_array_start(char chr)
+{
+	if (chr == '(') {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_array_stop(char chr)
+{
+	if (chr == ')') {
+		return true;
+	}
+
+	return false;
+}
+
+static inline bool is_number(char chr)
+{
+	if (isdigit(chr) ||
+	    (chr == '-') ||
+	    (chr == '+')) {
+		return true;
+	}
+
+	return false;
+}
+
+static inline void set_new_state(enum at_parser_state new_state)
+{
+	if ((state != SEARCH_NEXT_PARAM) &&
+	    (state != DETECT_TYPE)) {
+		prev_state = state;
+	}
+
+	state     = new_state;
+}
+
+static inline void reset_state(void)
+{
+	prev_state = DETECT_TYPE;
+	state      = DETECT_TYPE;
+}
+
+/*
+ * Internal function.
+ * Parameters cannot be null. String must be null terminated.
+ */
+static int at_parse_param(const char ** at_params_str,
+			  struct at_param_list * const list,
+			  const size_t max_params)
+{
+	int index = 0;
+	const char *str = *at_params_str;
+
+	reset_state();
+
+	while ((!is_terminated(*str)) &&
+	       (index < max_params)) {
+
+		str += isspace(*str);
```

Comment:
Ok, I change to at_params_get_space_count

Reply:
Changed it back to 
`
if (isspace(*str)) {
    str++;
}
`
I can only move the pointer one byte at the time to make the state machine work correctly in all cases. I can not use the at_params_trim_space, because then the state machine will loose some information about \r\n characters.